rclrs/src/node/mod.rs

```diff
@@ -109,7 +110,7 @@ impl Node {
         callback: F,
```

Comment:
Why would we have a redundant function like this? The current `Subscription` API can receive `Messages`, and both idiomatic and RMW-compatible messages are instances of that. Adding a second API that only supports the latter type is redundant and confuses only, imo. Uniformity is nice.

Reply:
Precisely for uniformity with the other bindings, no other bindings accept these so-call rmw types, all of the support idiomatic types. Adding a second API clears that confusion by explicitly signaling that it's specific to `ros2-rust`. Rust, very much like Python, and unlike C++, encourages explicitness vs implicitness. Explicitness is nice too.

Reply:
> through less-than-canonical APIs

The RMW-compatible message API is already non-canonical, because it does not have any counterpart in the other bindings. APIs are not only about names, but also semantics, the arguments they accept and the values they return. Having a distinctly different name emphasizes that it behaves differently from the other bindings, and that users should expect differences.

> Besides, there is also lots of stuff that is specific to rclcpp, and none of it is marked as such (topic statistics, generic pub/sub, intraprocess, etc.).

Because the goal is that eventually that functionality might be incorporated into other bindings. For example, `rclc` could potentially get support for intraprocess.

Anyway, I rather not waste more time on this, but it's important to get APIs right early on, because changing them later is far more difficult

Reply:
As for counterparts in other bindings, here's my perspective for posterity:
- Python is slow anyway, and usability of a raw C message (= our "RMW-compatible" message) would be absymal from a Python perspective, so of course the messages in that language are an idiomatic version that are converted to/from the C message.
- C and C++ have the privilege of being natively RMW-compatible, so they can be both idiomatic and maximally efficient. They are the counterpart to the RMW-compatible *and* the idiomatic messages.
- I assume other language bindings also use their own idiomatic message types that are converted to/from C messages. However, I believe none of the other languages are quite as performance-focused as Rust/C/C++.

Hence, Rust is kind of unique, in that its users expect maximum performance but it can't achieve that (yet) by serializing/deserializing its standard library types to/from CDR directly. So you could argue that it would be canonical to use RMW-compatible messages because that's what the other performance-focused language bindings do. Or you could argue that it would be canonical to use idiomatic messages because that's what all language bindings do.

I think the solution in this PR, to offer both, isn't too bad. And I placed the idiomatic messages in the `<pkg>::msg` namespace and the RMW-compatible ones in `<pkg>::msg::rmw`, because they can't be both in the same namespace, but I believe we shouldn't hide the RMW-compatible messages more than that if it's not technically necessary. 

I don't think the time spent discussing on this is wasted. Hope I'm not too stubborn! If you insist, I'll add the separate function. Or I can also poll the ROS + Rust people I know to get more input. But since you approved, I take it that it's okay to merge as it is now and potentially discuss it again later.

Reply:
> * I assume other language bindings also use their own idiomatic message types that are converted to/from C messages. However, I believe none of the other languages are quite as performance-focused as Rust/C/C++.

Ada reuses the C data layout in memory, so there is no overhead involved. The price is somewhat un-idiomatic use when sequences are involved that I hope to alleviate at some point in the future.